lib/libvfio-user.c

```diff
@@ -504,61 +505,54 @@ handle_dma_map_or_unmap(vfu_ctx_t *vfu_ctx, uint32_t size, bool map,
     }
 
     for (i = 0, fdi = 0; i < nr_dma_regions; i++) {
+        struct vfio_user_dma_region *region = &dma_regions[i];
+
+        vfu_log(vfu_ctx, LOG_DEBUG, "%s DMA region [%#lx-%#lx) offset=%#lx "
+                "prot=%#x flags=%#x", map ? "adding" : "removing",
+                region->addr, region->addr + region->size,
+                region->offset, region->prot, region->flags);
+
         if (map) {
             int fd = -1;
-            if (dma_regions[i].flags == VFIO_USER_F_DMA_REGION_MAPPABLE) {
+            if (region->flags == VFIO_USER_F_DMA_REGION_MAPPABLE) {
                 fd = consume_fd(fds, nr_fds, fdi++);
                 if (fd < 0) {
+                    vfu_log(vfu_ctx, LOG_ERR, "failed to add DMA region: "
+                            "mappable but fd not provided\n");
                     ret = fd;
                     break;
                 }
             }
 
-            ret = dma_controller_add_region(vfu_ctx->dma,
-                                            dma_regions[i].addr,
-                                            dma_regions[i].size,
-                                            fd,
-                                            dma_regions[i].offset,
-                                            dma_regions[i].prot);
+            ret = dma_controller_add_region(vfu_ctx->dma, (void *)region->addr,
+                                            region->size, fd, region->offset,
+                                            region->prot);
             if (ret < 0) {
                 if (fd != -1) {
                     close(fd);
                 }
-                vfu_log(vfu_ctx, LOG_INFO,
-                        "failed to add DMA region %#lx-%#lx offset=%#lx fd=%d: %s",
-                        dma_regions[i].addr,
-                        dma_regions[i].addr + dma_regions[i].size - 1,
-                        dma_regions[i].offset, fd,
+                vfu_log(vfu_ctx, LOG_ERR, "failed to add DMA region: %s",
```

Comment:
Don't we need to print the DMA region details in the error case?

Reply:
indeed, thanks.